File: cpp/api/src/base/impl/zmq/EventStreamSocketZmq.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "EventStreamSocketZmq.h"
#include "IdGenerator.h"
#include "Server.h"
#include "ContextZmq.h"
#include "Messages.h"

namespace cameo {

EventStreamSocketZmq::EventStreamSocketZmq() :
	m_context{nullptr} {

}

EventStreamSocketZmq::~EventStreamSocketZmq() {
	terminate();
}

void EventStreamSocketZmq::init(Context * context, const Endpoint& endpoint, RequestSocket * requestSocket) {

	m_context = dynamic_cast<ContextZmq *>(context);

	std::string cancelEndpoint {std::string("inproc://" + IdGenerator::newStringId())};

	// Create the sockets.
	m_cancelSocket = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(m_context->getContext(), zmq::socket_type::pub));
	m_cancelSocket->bind(cancelEndpoint);

	m_socket = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(m_context->getContext(), zmq::socket_type::sub));

	std::vector<std::string> streamList;
	streamList.push_back(message::Event::STATUS);
	streamList.push_back(message::Event::RESULT);
	streamList.push_back(message::Event::KEYVALUE);
	streamList.push_back(message::Event::CANCEL);

	for (std::vector<std::string>::const_iterator s = streamList.begin(); s != streamList.end(); ++s) {
		m_socket->set(zmq::sockopt::subscribe, *s);
	}

	m_socket->connect(endpoint.toString().c_str());
	m_socket->connect(cancelEndpoint);

	// Wait for the connection to be ready.
	// Poll subscriber.
	zmq_pollitem_t items[1];
	items[0].socket = static_cast<void *>(*(m_socket.get()));
	items[0].fd = 0;
	items[0].events = ZMQ_POLLIN;
	items[0].revents = 0;

	while (true) {
		try {
			requestSocket->request(createSyncRequest());
		}
		catch (const ConnectionTimeout& e) {
			// The server is not accessible.
		}

		// Wait for 100ms.
		int rc = zmq::poll(items, 1, std::chrono::milliseconds{100});
		if (rc != 0) {
			break;
		}
	}
}

void EventStreamSocketZmq::send(const std::string& data) {

	zmq::message_t messageData {data.c_str(), data.size()};
	m_socket->send(messageData, zmq::send_flags::none);
}

std::string EventStreamSocketZmq::receive(bool blocking) {

	// Use the message interface.
	zmq::message_t message;

	zmq::recv_flags flags = {(blocking ? zmq::recv_flags::none : zmq::recv_flags::dontwait)};
	if (m_socket->recv(message, flags).has_value()) {
		// The message exists.
		return std::string{message.data<char>(), message.size()};
	}

	return "";
}

void EventStreamSocketZmq::cancel() {

	if (m_cancelSocket.get() != nullptr) {
		std::string data {message::Event::CANCEL};

		zmq::message_t typePart {data.c_str(), data.length()};
		m_cancelSocket->send(typePart, zmq::send_flags::sndmore);

		zmq::message_t dataPart {data.c_str(), data.length()};
		m_cancelSocket->send(dataPart, zmq::send_flags::none);
	}
}

void EventStreamSocketZmq::terminate() {

	if (m_socket) {
		m_socket->close();
		m_socket.reset();
	}
}

}

File: java/api/src/main/java/eu/ill/cameo/api/base/impl/zmq/EventStreamSocketZmq.java
package eu.ill.cameo.api.base.impl.zmq;
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */



import java.util.concurrent.atomic.AtomicBoolean;

import org.json.simple.JSONObject;
import org.json.simple.parser.ParseException;

import eu.ill.cameo.api.base.ConnectionTimeout;
import eu.ill.cameo.api.base.Context;
import eu.ill.cameo.api.base.Event;
import eu.ill.cameo.api.base.IdGenerator;
import eu.ill.cameo.api.base.KeyEvent;
import eu.ill.cameo.api.base.RequestSocket;
import eu.ill.cameo.api.base.ResultEvent;
import eu.ill.cameo.api.base.StatusEvent;
import eu.ill.cameo.api.base.UnexpectedException;
import eu.ill.cameo.api.base.impl.EventStreamSocketImpl;
import eu.ill.cameo.com.Zmq;
import eu.ill.cameo.common.messages.JSON;
import eu.ill.cameo.common.messages.JSON.Parser;
import eu.ill.cameo.common.messages.Messages;
import eu.ill.cameo.common.strings.Endpoint;

public class EventStreamSocketZmq implements EventStreamSocketImpl {
	
	private Zmq.Context context;
	private Parser parser;
	private Zmq.Socket subscriberSocket;
	private Zmq.Socket cancelSocket;
	private AtomicBoolean canceled = new AtomicBoolean(false);
	
	public EventStreamSocketZmq() {
		super();
	}
	
	public void init(Context context, Endpoint endpoint, RequestSocket requestSocket, Parser parser) {
		
		this.context = ((ContextZmq)context).getContext();
		this.parser = parser;
				
		// Prepare our subscriber.
		Zmq.Socket subscriber = this.context.createSocket(Zmq.SUB);
		
		subscriber.connect(endpoint.toString());
		
		subscriber.subscribe(Messages.Event.STATUS);
		subscriber.subscribe(Messages.Event.RESULT);
		subscriber.subscribe(Messages.Event.KEYVALUE);
		
		String cancelEndpoint = "inproc://" + IdGenerator.newStringId();
		
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(Messages.Event.CANCEL);
		
		// polling to wait for connection
		Zmq.Poller poller = this.context.createPoller(1);
		poller.register(subscriber);
		
		while (true) {
			// The server returns a STATUS message that is used to synchronize the subscriber
			try {
				requestSocket.requestJSON(Messages.createSyncRequest());
			}
			catch (ConnectionTimeout e) {
				// Do nothing.
			}

			// Return after the first message received by the subscriber.
			poller.poll(100);
			if (poller.pollin(0)) {
				break;
			}
		}
		
		Zmq.Socket cancelPublisher = this.context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);
		
		this.subscriberSocket = subscriber;
		this.cancelSocket = cancelPublisher;
	}
	
	public Event receive() {
		
		String message = this.subscriberSocket.recvStr();
		Event event = null;
		
		// We can receive messages from the status publisher located in the server
		// as well as messages from the cancel publisher located in the same process.
		if (message.equals(Messages.Event.STATUS)) {
			
			byte[] statusMessage = this.subscriberSocket.recv();
			
			try {
				// Get the JSON object.
				JSONObject jsonObject = parser.parse(Messages.parseString(statusMessage));
				
				int id = JSON.getInt(jsonObject, Messages.StatusEvent.ID);
				String name = JSON.getString(jsonObject, Messages.StatusEvent.NAME);
				int state = JSON.getInt(jsonObject, Messages.StatusEvent.APPLICATION_STATE);
				int pastStates = JSON.getInt(jsonObject, Messages.StatusEvent.PAST_APPLICATION_STATES);
								
				if (jsonObject.containsKey(Messages.StatusEvent.EXIT_CODE)) {
					int exitCode = JSON.getInt(jsonObject, Messages.StatusEvent.EXIT_CODE);
					event = new StatusEvent(id, name, state, pastStates, exitCode);
				}
				else {
					event = new StatusEvent(id, name, state, pastStates);	
				}
			}
			catch (ParseException e) {
				throw new UnexpectedException("Cannot parse response");
			}
		}
		else if (message.equals(Messages.Event.RESULT)) {
				
			byte[] resultMessage = this.subscriberSocket.recv();
			
			try {
				// Get the JSON object.
				JSONObject jsonObject = parser.parse(Messages.parseString(resultMessage));
				
				int id = JSON.getInt(jsonObject, Messages.ResultEvent.ID);
				String name = JSON.getString(jsonObject, Messages.ResultEvent.NAME);
				
				// Get the next message to get the data.
				byte[] data = this.subscriberSocket.recv();
				
				event = new ResultEvent(id, name, data);
			}
			catch (ParseException e) {
				throw new UnexpectedException("Cannot parse response");
			}
		}
		else if (message.equals(Messages.Event.KEYVALUE)) {
			
			byte[] keyValueMessage = this.subscriberSocket.recv();
			
			try {
				// Get the JSON object.
				JSONObject jsonObject = parser.parse(Messages.parseString(keyValueMessage));
				
				int id = JSON.getInt(jsonObject, Messages.KeyEvent.ID);
				String name = JSON.getString(jsonObject, Messages.KeyEvent.NAME);
				long status = JSON.getLong(jsonObject, Messages.KeyEvent.STATUS);
				String key = JSON.getString(jsonObject, Messages.KeyEvent.KEY);
				String value = JSON.getString(jsonObject, Messages.KeyEvent.VALUE);
				
				if (status == Messages.STORE_KEY_VALUE) {
					event = new KeyEvent(id, name, KeyEvent.Status.STORED, key, value);
				}
				else {
					event = new KeyEvent(id, name, KeyEvent.Status.REMOVED, key, value);
				}
			}
			catch (ParseException e) {
				throw new UnexpectedException("Cannot parse response");
			}
		}
		else if (message.equals(Messages.Event.CANCEL)) {
			canceled.set(true);
			return null;
		}
	
		return event;
	}

	public void cancel() {
		cancelSocket.sendMore(Messages.Event.CANCEL);
		cancelSocket.send(Messages.Event.CANCEL);
	}
	
	public boolean isCanceled() {
		return canceled.get();
	}
		
	public void terminate() {
		context.destroySocket(subscriberSocket);
	}
}